src/spawn.c

```diff
@@ -562,26 +564,99 @@ static gboolean spawn_async_with_pipes(const gchar *working_directory, const gch
 	g_message("full spawn command line: %s", command->str);
 #endif
 
-	while (envp && *envp)
+	while (envp && *envp && success)
 	{
-		g_array_append_vals(environment, *envp, strlen(*envp) + 1);
+		gsize locale_entry_len;
+		gchar *locale_entry;
+
+		if (g_utf8_validate(*envp, -1, NULL))
+		{
+			/* TODO: better error message */
+			locale_entry = g_locale_from_utf8(*envp, -1, NULL, &locale_entry_len, error);
+		}
+		else
+		{
+			locale_entry_len = strlen(*envp);
+			locale_entry = g_memdup(*envp, locale_entry_len + 1);
+		}
+
+		if (! locale_entry)
+			success = FALSE;
+		else
+		{
+			/* copy the entry, including NUL terminator */
+			g_array_append_vals(environment, locale_entry, locale_entry_len + 1);
+			g_free(locale_entry);
+		}
+
 		envp++;
 	}
 
-	failure = spawn_create_process_with_pipes(command->str, working_directory,
-		envp ? environment->data : NULL, child_pid, stdin_fd, stdout_fd, stderr_fd);
+	/* convert working directory into locale encoding */
+	if (success && working_directory)
+	{
+		if (g_utf8_validate(working_directory, -1, NULL))
+		{
+			GError *gerror = NULL;
 
-	g_string_free(command, TRUE);
-	g_array_free(environment, TRUE);
+			locale_working_directory = g_locale_from_utf8(working_directory, -1, NULL, NULL, &gerror);
+			if (! locale_working_directory)
+			{
+				/* TODO use the code below post-1.28 as it introduces a new string
```

Comment:
comment is now outdated -- although it mostly still applies… bad us not to work on that earlier!^^